### PR TITLE
Organize and refactor nav2_util client code

### PR DESCRIPTION
**Pull request by bpwilcox**
This PR organizes the service clients in `nav2_util` and creates a simple subscriber class.

## Description of contribution in a few bullet points
- create include directory `nav2_clients`
- create service clients for `ClearAroundCostmapServiceClient` and `ClearExceptCostmapServiceClient`
- adds `clearAroundCostmapServiceRequest` and `clearExceptCostmapServiceRequest` to `NavigateToPoseBehaviorTree`
- creates `SimpleSubscriber` of which now `CostmapSubscriber` and `FootprintSubscriber` derive
---

## Future work that may be required in bullet points
- Organize dependencies of client libraries

### CODE REVIEW

**Comment by crdelsey** [comment]
@bpwilcox Please rebase before review

File: nav2_util/include/nav2_util/simple_subscriber.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        qos, std::bind(&SimpleSubscriber::subscriberCallback, this, std::placeholders::_1));
+  }
+
+  std::shared_ptr<SubscriberT> getMsg()
+  {
+    if (!received_) {
+      std::string error(topic_name_ + " not yet received");
+      throw std::runtime_error(error);
+    }
+    return msg_;
```
Comment (mjeronimo) [inline nav2_util/include/nav2_util/simple_subscriber.hpp]: getMsg() could be called from a different thread than the subscriberCallback. The member variables (received_, msg_) should be protected from simultaneous access from different threads.

File: nav2_costmap_2d/include/nav2_costmap_2d/footprint_subscriber.hpp [hunk 2]
```diff
@@ -20,43 +20,27 @@
...
+#include "nav2_util/simple_subscriber.hpp"
 
 namespace nav2_costmap_2d
 {
 
-class FootprintSubscriber
+class FootprintSubscriber : public nav2_util::SimpleSubscriber<geometry_msgs::msg::PolygonStamped>
 {
 public:
   FootprintSubscriber(
```
Comment (mjeronimo) [inline nav2_costmap_2d/include/nav2_costmap_2d/footprint_subscriber.hpp]: FYI: In ROS2, the pattern of which constructors to provide isn't well established yet. However, the prevalent pattern seems to be providing the interface-based constructor and the rclcpp::Node-based constructor. They seem to be providing lifecycle support by requiring a lifecycle node to use the interface pointers. See tf2_ros/message_filter.h, for example. 

There are a few options for us to choose from:

1. Separate constructors for: rclcpp::Node, rclcpp_lifecycle::LifecycleNode, node interfaces (all three options)
2. Separate constructors for: rclcpp::Node, node interfaces (this is what many ROS2 classes are using)
3. Separate constructors for: rclcpp:Node, rclcpp_lifecycle::LifecycleNode (what's used here)

Previously, Carlos used #1 (with the old robot class). For consistency, we should standardize on an approach when implementing helper/support classes that work on behalf of a node.

**Comment by mkhansen-intel** [comment]
Is this PR still active? It looks like it needs rebasing. If it's not active let's close it, otherwise please rebase and update so we can review it.

File: nav2_util/include/nav2_util/simple_subscriber.hpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+{
+
+template<class SubscriberT>
+class SimpleSubscriber
+{
+public:
+  SimpleSubscriber(
+    nav2_util::LifecycleNode::SharedPtr node,
+    std::string & topic_name,
+    rclcpp::QoS & qos)
```
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/simple_subscriber.hpp]: No default for default QoS?

File: nav2_util/include/nav2_util/nav2_clients/clear_except_costmap_service_client.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2019 Intel Corporation
```
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/nav2_clients/clear_except_costmap_service_client.hpp]: I don't like the idea of adding additional utilities no application utilizes. I think keeping these on the back burner until there's something using it is best

**Comment by mkhansen-intel** [comment]
This is low priority and Brian isn't working on it right now, so closing. Can re-open later.

**Pull request closed without merging**